views: endpoint for item state by pid

kpsherva opened the pull request:
* Exposed endpoint for item circulation_state (closes #70)

* normalized api links

kpsherva committed:
views: endpoint for item state by pid

* Exposed endpoint for item circulation_state (closes #70)

* normalized api links

zzacharo commented on invenio_circulation/views.py:
```diff
@@ -137,3 +141,65 @@ def post(self, pid, record, action, **kwargs):
+    }
+
+    from invenio_circulation.links import loan_links_factory
+    loan_request = StateResource.as_view(
```
You forgot to rename this one :)

zzacharo commented on invenio_circulation/views.py:
```diff
@@ -137,3 +141,65 @@ def post(self, pid, record, action, **kwargs):
+    return blueprint
+
+
+class StateResource(ContentNegotiatedMethodView):
```
Maybe be explicit `ItemStateResource`?

ntarocco replied:
Yes, or maybe `ItemCirculationState`? Since it is not the state of the item, it is more the state of circulation for this item.
Same for `view_name`...

zzacharo commented on invenio_circulation/views.py:
```diff
@@ -137,3 +141,65 @@ def post(self, pid, record, action, **kwargs):
+    def get(self, *args, **kwargs):
+        """Handle GET request for item state."""
+        loans = list(LoansSearch.search_loans_by_pid(
+            item_pid=kwargs.get('pid_value', None),
```
What if item_pid is not passed? Do we want to list all `ITEM_ON_LOAN` loans?

ntarocco replied:
Agree:
1) add a test in case of "url malformed"
2) don't we want to return the current loan, if any, without filtering by state? Or did u have something else in mind?

kpsherva replied:
as I see `item_pid` is string so either it won't match anything for `Loan`, or it will return 404 in case of empty string

zzacharo commented on invenio_circulation/views.py:
```diff
@@ -137,3 +141,65 @@ def post(self, pid, record, action, **kwargs):
+        for key, value in ctx.items():
+            setattr(self, key, value)
+
+    def get(self, *args, **kwargs):
```
Probably we should expose a permission factory which be default allows all but someone can override it in his/her module(e.g invenio-app-ils).

ntarocco replied:
Yes, this is related to the bug I have opened: https://github.com/inveniosoftware/invenio-circulation/issues/72

If you want to fix it, and do the same for the other view too, here an example:
https://github.com/inveniosoftware/invenio-app-ils/blob/wip-rest-patron-loan-history/invenio_app_ils/circulation/views.py#L77

kpsherva replied:
I've added the permissions, by default it allows all

ntarocco commented on invenio_circulation/config.py:
```diff
@@ -131,8 +131,8 @@
         },
-        list_route='/circulation/loan/',
-        item_route='/circulation/loan/<{0}:pid_value>'.format(_Loan_PID),
+        list_route='/circulation/loans/',
```
thanks!

ntarocco commented on invenio_circulation/views.py:
```diff
@@ -137,3 +141,65 @@ def post(self, pid, record, action, **kwargs):
+
+    rec_serializers = {
+        "application/json": (
+            "invenio_records_rest.serializers" ":json_v1_response"
```
We should find a way to "hook in" our serializer because of https://github.com/inveniosoftware/invenio-app-ils/issues/25 instead of hardcoding, but maybe we can handle when working on that.

kpsherva replied:
is it for me to resolve now or put it into the issues?

ntarocco replied:
I think it can be fixed while working on inveniosoftware/invenio-app-ils#25

ntarocco commented on tests/test_api_calls.py:
```diff
@@ -132,3 +132,30 @@ def test_api_loans_links_factory(app, db, json_headers, params,
+    multiple_loans_pid = "item_multiple_pending_on_loan_7"
+
+    with app.test_client() as client:
+        # one pending load
```
What about splitting in 3 different tests?
Something like (better naming probably):
`test_get_item_circulation_status_when_item_not_found` (should assert response code == 404?)
`test_get_item_circulation_status_when_none`
`test_get_item_circulation_status_when_exists`

Could you please test also the response code?

ntarocco commented on invenio_circulation/views.py:
```diff
@@ -137,3 +141,65 @@ def post(self, pid, record, action, **kwargs):
+        ctx=dict(links_factory=loan_links_factory),
+    )
+
+    url = 'circulation/items/<pid_value>/status'
```
Just thinking now, since the part of the url `/circulation` is configurable by the REST config and it could be changed when configuring the module, maybe we should have this also using the same path.
What about trying to get `current_app.config['CIRCULATION_REST_ENDPOINTS']['list_route']` and try to guess the `/circulation` part, but it is probably difficult, I don't know if you have good ideas.
```

ntarocco commented on invenio_circulation/views.py:
```diff
@@ -137,3 +141,65 @@ def post(self, pid, record, action, **kwargs):
+def build_blueprint_with_state(app):
+    """Item circulation state blueprint."""
+    blueprint = Blueprint(
+        'invenio_circulation_state',
```
Same about naming, similar to the others,`invenio_circulation_item_circulation_state`? (but the duplication of `circulation` is ugly :()

kpsherva committed:
permissions: add default permission factory

* views protected by allow all permission
 (option to overwrite the access to read the loan object)

* protect LoanAction views (closes #72)

* tests for different cases of users permissions

zzacharo commented on tests/conftest.py:
```diff
@@ -151,3 +160,48 @@ def mock_is_item_available():
         yield mock_is_item_available
+
+
+@pytest.fixture()
```
This is not needed as it's used as a function not as a fixture :)

kpsherva replied:
done :)

zzacharo commented on tests/test_api_perms.py:
```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
```
Better name the file with the whole name `test_api_permissions.py` to be more readable?

zzacharo commented:
I don't know if my comment is seen but I proposed to rename the file `tests/test_api_perms.py` to `tests/test_api_permissions.py`

ntarocco commented:
Is this PR to be re-reviewed?

kpsherva commented:
@ntarocco yes, you can re-review it, I just had one question there above about the serializer.

ntarocco commented on invenio_circulation/views.py:
```diff
@@ -115,6 +121,7 @@ def __init__(self, serializers, ctx, *args, **kwargs):
         for key, value in ctx.items():
             setattr(self, key, value)
 
+    @need_permissions('loan-read-access')
```
Minor: I would switch the order of the decorator to have first `need_permissions` check, and if it fail it does not do the db call for the `pass_record`, WDTY?

ntarocco replied:
Thanks for this fix too! 👍

ntarocco commented on requirements-devel.txt:
```diff
@@ -6,6 +6,7 @@
 # Invenio-Circulation is free software; you can redistribute it and/or modify it
 # under the terms of the MIT License; see LICENSE file for more details.
 
+-e git+https://github.com/inveniosoftware/invenio-access#egg=invenio-access
```
can you please add `invenio-access` in the setup.py too?

kpsherva committed:
views: check if multiple active loans on item

* Exposing item loan states considered as active

* Fix package requirements in setup

* Test against more than one loan is active on an item

kpsherva commented:
I have updated the endpoint with active states of the loan, and I've added the checking if there is more than one active loan on the item, it is ready to be checked

ntarocco reviewed (approved):
LGTM!
